src/utils.ts

```diff
@@ -0,0 +1,78 @@
+'use strict'
+
+import { Snippet } from './extension'
+
+interface SnippetGrammar {
+  tokens: Token[]
+}
+
+interface Token {
+  snippet: Pattern
+  variable: Pattern
+}
+
+interface Pattern {
+  re: RegExp
+  replacement: string
+}
+
+export const TABSTOP: Token = {
+  snippet: {
+    re: /\$(\d)/g,
+    replacement: 'TABSTOP_$1',
+  },
+  variable: {
+    re: /TABSTOP_(\d)/g,
+    replacement: '$$$1',
+  },
+}
+
+export const PLACEHOLDER: Token = {
+  snippet: {
+    re: /\$\{(\d):(\w+)\}/g, // tabstop : default value
+    replacement: 'PLACEHOLDER_$1_$2',
+  },
+  variable: {
+    re: /PLACEHOLDER_(\d)_(\w+)/g,
+    replacement: '${$1:$2}',
+  },
+}
+
+export function parseSnippets(
+  snippets: { [name: string]: Snippet },
+  from: string,
+  grammar: SnippetGrammar
+) {
+  const _snippets = Object.assign({}, snippets)
+
+  Object.keys(_snippets).map(name => {
+    _snippets[name].body = parse(_snippets[name].body, from, grammar)
+  })
```

Comment:
This is why I proposed `Object.keys` + `reduce`, because you could to this in one step.
```javascript
const obj = { hello: 'react', hi: 'redux' }

const result = Object.keys(obj).reduce((acc, key) => {
  // do stuff
  acc[key] = obj[key]
  return acc
}, {}) <-- initial value is important here
```